oauth/api/router.py

```diff
@@ -0,0 +1,167 @@
+from fastapi.security import HTTPBasic
+from sqlalchemy.orm import Session
+
+import oauth.api.endpoints as endpoints
+from oauth.api.exceptions import (
+    AuthenticationException,
+    ClientNotFoundException,
+    InvalidScopeException,
+)
+from oauth.api.models import (
+    AccessToken,
+    ClientCreatedResponse,
+    OAuth2ClientCredentialsRequestForm,
+)
+from oauth.api.utils import verify_oauth_client
+from oauth.database.models import ClientDetail
+from oauth.scopes import (
+    CLIENT_CREATE,
+    CLIENT_DELETE,
+    CLIENT_READ,
+    CLIENT_UPDATE,
+    SCOPE_READ,
+    SCOPES,
+)
+
+logger = getLogger(__name__)
+router = APIRouter(prefix=endpoints.OAUTH_PREFIX, tags=["OAuth"])
+
+
+@router.post(endpoints.TOKEN, response_model=AccessToken)
+async def acquire_access_token(
+    request: Request,
+    form_data: OAuth2ClientCredentialsRequestForm = Depends(),
+    db: Session = Depends(),
+) -> AccessToken:
+    """
+    Given a set of credentials, returns an access token.
+    """
+
+    basic_credentials = await HTTPBasic(auto_error=False)(request)
+
+    if form_data.client_id and form_data.client_secret:
+        client_id = form_data.client_id
+        client_secret = form_data.client_secret
+    elif basic_credentials:
+        client_id = basic_credentials.username
+        client_secret = basic_credentials.password
+    else:
+        raise AuthenticationException()
+
+    client_detail = ClientDetail.get(db, client_id=client_id)
+
+    if client_detail is None:
+        raise AuthenticationException()
+
+    if not client_detail.credentials_valid(client_secret):
+        raise AuthenticationException()
+
+    logger.info("Creating access token for client with ID '%s'", client_id)
+    return AccessToken(access_token=client_detail.create_access_code_jwe())
+
+
+@router.post(
+    endpoints.CLIENT,
+    dependencies=[Security(verify_oauth_client, scopes=[CLIENT_CREATE])],
+    response_model=ClientCreatedResponse,
+    status_code=status.HTTP_201_CREATED,
+)
+def create_client(
+    *,
+    db: Session = Depends(),
+    scopes: List[str] = Body([]),
+) -> ClientCreatedResponse:
+    """
+    Creates a new client and returns the credentials.
+    """
+
+    for scope in scopes:
+        if not scope in SCOPES:
+            raise InvalidScopeException(scope)
+
+    client, secret = ClientDetail.create_client_and_secret(db, scopes)
+    logger.info("Created new client with ID '%s'", client.id)
+
+    return ClientCreatedResponse(client_id=client.id, client_secret=secret)
+
+
+@router.delete(
+    endpoints.CLIENT_BY_ID,
+    dependencies=[Security(verify_oauth_client, scopes=[CLIENT_DELETE])],
+)
+def delete_client(client_id: str, db: Session = Depends()) -> None:
+    """
+    Deletes the client associated with the client_id.
+    """
+
+    client = ClientDetail.get(db, client_id=client_id)
+    if not client:
+        raise ClientNotFoundException(client_id)
```

Comment:
This is a divergence from the current behavior in fidesops, and represents a significant change to the contract. Fidesops currently no-ops when a client cannot be found. This felt like better handling to me, but it's unclear if the no-op is depended upon by API consumers.

Reply:
That's a safe change to make, go ahead with it. A 404 would be a very reasonable thing to do; failing silently will invite errors

Reply:
I agree with @pattisdr here, this was the whole reason we left it ambiguous. Otherwise you're vulnerable to brute force ID guessing

Reply:
@pattisdr @seanpreston Excellent point, and I agree. I've changed this endpoint to always respond with a `204 No Content` and an empty response payload. The logs will only indicate that a client resource was deleted if a client was actually found to delete. I think this is safe because [presumably] if you have access to the server logs then you also have access to the valid records?